src/core/agent/AgentService.ts

```diff
@@ -272,10 +283,93 @@ export default class AgentService {
         }
         return dump
     }
+
+    async getCapabilities(token: string) {
+        if (token == this.#adminCredential) {
+            return [ALL_CAPABILITY]
+        }
+        
+        if (token === '') {
+            return [AUTH_CAPABILITY]
+        }
+
+        const key = this.getSigningKey()
+        let keyEncoder = new KeyEncoder('secp256k1')
+        const pemPublicKey = keyEncoder.encodePublic(key.publicKey, 'raw', 'pem')
+
+        const pubKeyObj = crypto.createPublicKey(pemPublicKey)
+        const { payload } = await jose.jwtVerify(token, pubKeyObj)
+
+        return payload.capabilities
+    }
+    
+    requestAuth(appName: string, appDesc: string, appUrl: string, capabilities: string) {
+        let requestId = uuidv4()
+        let authExtended = {
+            requestId,
+            auth: {
+                appName,
+                appDesc,
+                appUrl,
+                capabilities: JSON.parse(capabilities),
+            } as AuthInfo,
+        } as AuthInfoExtended
+        
+        this.#pubsub.publish(
+            PubSubInstance.EXCEPTION_OCCURRED_TOPIC,
+            {
+                title: "Request to authenticate application",
+                message: `${appName} is waiting for authentication, go to ad4min for more information.`,
+                type: ExceptionType.RequestAuth,
+                addon: JSON.stringify(authExtended),
+            } as ExceptionInfo
+        )
+
+        return requestId
+    }
+
+    permitAuth(adjustedAuth: string, capabilities: Capability[]) {
+        console.log("admin user capabilities: ", capabilities)
+        console.log("auth info: ", adjustedAuth)
+
+        let { requestId, auth }: AuthInfoExtended = JSON.parse(adjustedAuth)
+        let rand = genAuthRand()
+        this.#tokens.set(genAuthKey(requestId, rand), auth)
+        
```

Comment:
So basically we can move this code into `requestAuth()` (which I think should be called `requestCapability()`.

Reply:
Authenticate is a process, capability is part of the resource of such process. I tried to rename `requestAuth` to `requestCapability`, and `permitAuth` to `permitCapability`, but it feels not straight, since we have `requestCapabilities` in the params of `requestCapability`.